### PR TITLE
Feature/popover validation

### LINKED ISSUE #60: Validate Popover Form

### PR DESCRIPTION
**Pull request by charkour**
This feature adds form validation to the popover. It requires prefix, number, and section are added. This also makes use of a Form context provider which cuts down on repeated LOC. That's why the additions and deletions are high. 

To test:
- try to click submit on a pop over to see the errors and then try to fix them or pass in bad values. 
- test clicking a created event and then seeing that the data fill properly

Finally, I suggest we make the popover into a full-page in the future because there are so many options that need to be adjusted and we should have a simplified version of editing that is a popover. I'm basing a lot of the design decisions on Google Calendar.

### CODE REVIEW

File: client-course-schedulizer/src/utilities/services/validation/popover.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  number: number().typeError("number must be a number").required().positive().integer(),
+  prefix: string().required().uppercase(),
+  roomCapacity: number().typeError("global max must be a number").positive().integer(),
+  section: string().required().uppercase(),
+  studentHours: number()
+    .typeError("student hours must be a number")
+    .required()
+    .positive()
+    .integer(),
+});
```
Comment (bryantgeorge) [inline client-course-schedulizer/src/utilities/services/validation/popover.ts:105]: For the validation:
1. Having the error messages list fields in camel case is slightly strange, but not a big deal.
2. `facultyHours` and `studentHours` should be allowed to be decimals (like 1.5).
3. I don't think `location`, `days`, and `duration` should be required due to online asynchronous classes.
4. Similarly with `instructor`, I wonder if it shouldn't be required to allow for schedules where things aren't fully planned out yet?

File: client-course-schedulizer/src/utilities/services/addSectionService.ts [hunk 2]
```diff
@@ -142,20 +203,23 @@ export const mapInputToInternalTypes = (data: SectionInput) => {
...
         },
         startTime: startTimeCase(data.startTime),
       },
     ],
     semesterLength: semesterType,
     studentHours: Number(data.studentHours),
     term: data.term,
-    year: "2021-2022",
+    // TODO: there currently isn't a way to get the year for this Section.
+    year: `${"hello"}`,
```
Comment (bryantgeorge) [inline client-course-schedulizer/src/utilities/services/addSectionService.ts]: For testing with dept chairs soon, should we have it assume 2021? (Also I noticed in the `Course Section Enrollment.csv` that they list 2020 as the academic year for all classes, even if the class is in spring when it's 2021)

**Comment by bryantgeorge** [comment]
I noticed that there's an `x` button on the time slot which allows you to clear the time (or by pressing backspace) which can lead to meetings without time, is this intentional? (This might be out of the scope of this PR because I assume this is how it always worked)

image

**Review by bryantgeorge — CHANGES_REQUESTED** [review]
I like how convenient this is, good work! I agree that making it a separate page later will be good because of how big it is. Also I might have an idea for fixing the multiday classes bug, so I can look at that in #65.

File: client-course-schedulizer/src/utilities/services/validation/popover.ts (hunk 1 above)
Comment (charkour) [reply client-course-schedulizer/src/utilities/services/validation/popover.ts:105]: Yeah, those are good points. I will attempt to fix those.

**Comment by charkour** [comment]
@Syobnaf, Thanks for the comments! I'll address these over the break since I don't think there is a rush to add validation.

**Comment by charkour** [comment]
> I noticed that there's an `x` button on the time slot which allows you to clear the time (or by pressing backspace) which can lead to meetings without time, is this intentional? (This might be out of the scope of this PR because I assume this is how it always worked)
> 
> image

I couldn't reproduce this on Chrome MacOS. Maybe it is browser/OS specific?

*Edit*
I was able to reproduce it. I'll look into it but I think it'll be another PR

File: client-course-schedulizer/src/utilities/services/addSectionService.ts (hunk 2 above)
Comment (charkour) [reply client-course-schedulizer/src/utilities/services/addSectionService.ts]: I like your fix in pr #69 . Thanks

File: client-course-schedulizer/src/utilities/services/validation/popover.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    .test("is-decimal", "invalid decimal", decimalRegex)
+    .transform(emptyStringToNull)
+    .nullable(),
+  year: number()
+    .typeError("year must be a number")
+    .positive()
+    .integer()
+    .transform(emptyStringToNull)
+    .nullable(),
+});
```
Comment (bryantgeorge) [inline client-course-schedulizer/src/utilities/services/validation/popover.ts:105]: I think we may want to add checks for the `used` and `day10Used` fields like:
```ts
  used: number()
    .typeError("used must be a number")
    .positive()
    .integer()
    .transform(emptyStringToNull)
    .nullable(),
  day10Used: number()
    .typeError("day 10 used must be a number")
    .positive()
    .integer()
    .transform(emptyStringToNull)
    .nullable(),
```

File: client-course-schedulizer/src/utilities/services/validation/popover.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  instructor: string(),
+  localMax: number()
+    .typeError("global max must be a number")
+    .positive()
+    .integer()
+    .transform(emptyStringToNull)
+    .nullable(),
+  location: string(),
+  name: string(),
+  number: number().typeError("number must be a number").required().positive().integer(),
```
Comment (bryantgeorge) [inline client-course-schedulizer/src/utilities/services/validation/popover.ts]: Sorry I didn't consider this earlier, but should we allow a string here for classes like `SPAN-W80-A` where the number is a string like `W80`? See the full schedule for other examples like this.

**Review by bryantgeorge — CHANGES_REQUESTED** [review]

File: client-course-schedulizer/src/utilities/services/validation/popover.ts (hunk 4 above)
Comment (charkour) [reply client-course-schedulizer/src/utilities/services/validation/popover.ts]: Ah, good catch! I forgot about that too. Thanks. We can just allow a string.

File: client-course-schedulizer/src/utilities/services/validation/popover.ts (hunk 3 above)
Comment (charkour) [reply client-course-schedulizer/src/utilities/services/validation/popover.ts:105]: Sounds good. Thanks!

**Review by bryantgeorge — APPROVED** [review]
LGTM

**Pull request merged**